### PR TITLE
add support for formatting files in subdirectories of `tests`

### LINKED ISSUE #1820: `cargo fmt` doesn't format files in sub-directories in `tests`
I have a big integration test suite with sub-directories inside of the `tests` directory; running `cargo fmt` doesn't format these files.

### PR DESCRIPTION
**Pull request by calebcartwright**
Resolves #1820

The idea of the approach is that if at least one of the packages that's going to be formatted contains a `test` target, then look for files within subdirectories of the `tests` directory for that package. If there were any found, then those are passed (along with the typical `src_path` for each of the package targets) to rustfmt.

I've tested it with a mix of repositories, format strategies, etc., including on clippy to validate. AFAICT it is working as desired. 

Opening this in draft to see whether folks think this might be a viable approach, if so I'll add some tests and do a bit of cleanup. It's not perfect, but my hope is that it's better than _not_ having any support for these files at all.

To test this with clippy (or any other repo that has nested files under `tests` as described in #1820), create a formatting error in one of those nested files (I butchered `tests/ui-toml/bad_toml/conf_bad_toml.rs` for my below examples, make sure you add `tests/ui/crashes/ice-3891.rs` to the rustfmt.toml `ignore` config)

Without the new flags, no checking of nested files and no isses
```sh
$ cargo +nightly fmt -- --check 
```

With the new flags:
```sh
$ cargo +nightly fmt --include-nested-test-files -- --check
Diff in rust-clippy/tests/ui-toml/bad_toml/conf_bad_toml.rs at line 1:
 // error-pattern: error reading Clippy's configuration file

-fn main()           {
-
-        }
+fn main() {}
```

I've made this an opt-in feature via a new flag for `cargo fmt` (`--include-nested-test-files`) given that this has the potential to introduce new formatting issues for folks (checking/formatting files that may have never been checked). We could potentially invert that (perhaps with 2.0?) to make this the default behavior with an opt-_out_ flag as well.

~~I also added a `cargo fmt` option to allow for listed nested files under `tests` to be ignored. The [ignore](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#ignore) config option also works for excluding files in subdirs within `tests`, but I was having an issue with one of the [ice files in the clippy repo](https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/crashes/ice-3891.rs) that was still happening even with that file set in the `rustfmt.toml` ignore setting 🤷‍♂~~

*Edit*: #3782 eliminated the need for this duplicative ignore cli arg

### CODE REVIEW

File: src/cargo-fmt/main.rs [hunk 1]
```diff
@@ -57,6 +58,14 @@ pub struct Opts {
...
     #[structopt(long = "all")]
     format_all: bool,
+
+    /// Format files located within subdirectories of `tests` directory
+    #[structopt(long = "include-nested-test-files")]
+    include_nested_test_files: bool,
+
+    /// Files within subdirectories of `tests` directory to ignore
+    #[structopt(long = "ignored-nested-test-files")]
+    ignored_nested_test_files: Vec<String>,
```
Comment (calebcartwright) [inline src/cargo-fmt/main.rs]: Not convinced these are the best names for these opts so please let me know if anyone has any suggestions (they could certainly use a short version too!)

File: src/cargo-fmt/main.rs [hunk 2]
```diff
@@ -398,19 +475,27 @@ fn get_targets_recursive(
...
                 .find(|p| p.name == dependency.name && p.source.is_none());
-            let manifest_path = if dependency_package.is_some() {
-                PathBuf::from(&dependency_package.unwrap().manifest_path)
-            } else {
-                let mut package_manifest_path = PathBuf::from(&package.manifest_path);
-                package_manifest_path.pop();
-                package_manifest_path.push(&dependency.name);
-                package_manifest_path.push("Cargo.toml");
-                package_manifest_path
+
```
Comment (calebcartwright) [inline src/cargo-fmt/main.rs]: This change isn't directly related, but figured I'd update this while I was in here

File: src/cargo-fmt/main.rs [hunk 3]
```diff
@@ -446,10 +537,99 @@ fn get_targets_with_hitlist(
...
+) -> Vec<PathBuf> {
+    let mut files = vec![];
+    if path.is_dir()
+        && !ignore_set
+            .matched_path_or_any_parents(&path, true)
+            .is_ignore()
+    {
+        for entry in fs::read_dir(path).expect(&format!(
+            "couldn't read directory {}",
+            path.to_str().unwrap()
```
Comment (calebcartwright) [inline src/cargo-fmt/main.rs]: This function was pretty much a copy/pasta from [here](https://github.com/rust-lang/rustfmt/blob/master/src/test/mod.rs#L91). I left the `.expect(`'s in there for now. Can clean them up if folks think the overall approach in this PR might be viable

**Comment by calebcartwright** [comment]
> Do you mean that formatting the file causes rustfmt to crash?

Correct. Without excluding that file from being passed to rustfmt in the first place, I always get the below error:
```sh
error: invalid suffix `x` for integer literal
 --> rust-clippy/tests/ui/crashes/ice-3891.rs:3:5
  |
3 |     1x;
  |     ^^ invalid suffix `x`
  |
  = help: the suffix must be one of the integral types (`u32`, `isize`, etc)
```

It looks like they were explicitly ignoring that file over in clippy too via their workaround

https://github.com/rust-lang/rust-clippy/blob/master/clippy_dev/src/fmt.rs#L51

**Comment by topecongiro** [comment]
Ok, looks like rustfmt is panicking inside the parser while trying to parse the file with invalid syntax.

The file is being parsed even though it is specified in `ignore` config option because we are filtering files with `ignore` after the module resolution, which requires parsing the file.

https://github.com/rust-lang/rustfmt/blob/1ded995ee79c44069b7832e5aba00162fc2113fa/src/formatting.rs#L101-L109

We should probably pass `ignore_path_set` to `ModResolver` and not parse the file that matches to the ignore pattern.

**Comment by calebcartwright** [comment]
> The file is being parsed even though it is specified in `ignore` config option because we are filtering files with `ignore` after the module resolution, which requires parsing the file.

Ahh that makes sense, thank you so much! I spent a lot of time trying to figure out if I had a typo in the `ignore` setting 😄 

> We should probably pass ignore_path_set to ModResolver and not parse the file that matches to the ignore pattern.

👍 IMO that sounds like a useful change even independent of this PR. I can create a new issue to track that (feel free to assign to me), and will open up a separate PR for that change. In parallel I'll update this PR to remove the `ignored-nested-testfiles` option over here (which will help simplify things) if that sounds good to you?

**Comment by topecongiro** [comment]
> 👍 IMO that sounds like a useful change even independent of this PR. I can create a new issue to track that (feel free to assign to me), and will open up a separate PR for that change. In parallel I'll update this PR to remove the ignored-nested-test-files option over here (which will help simplify things) if that sounds good to you?

SGTM. Thanks!

**Comment by calebcartwright** [comment]
CI failures seem unrelated AFAICT. Current master branch is also failing to compile against current nightly. Both this branch and master compile with passing tests against `nightly-2019-10-13`

File: src/cargo-fmt/main.rs [hunk 4]
```diff
@@ -748,4 +861,197 @@ mod cargo_fmt_tests {
             );
         }
     }
+
+    mod get_nested_integration_test_files_tests {
```
Comment (calebcartwright) [inline src/cargo-fmt/main.rs]: As this new behavior is on the `cargo fmt` side determining the correct set of files to pass to `rustfmt`, I wanted to ensure we had some automated tests. I decided to create some mini-sample projects under the `tests` directory and then adding some tests here against the functions that are determining the targets/files.

File: src/cargo-fmt/main.rs [hunk 5]
```diff
@@ -748,4 +861,197 @@ mod cargo_fmt_tests {
...
+            let path = PathBuf::from(src_path);
+            let canonicalized = fs::canonicalize(&path).unwrap_or(path);
+            Target {
+                path: canonicalized,
+                kind: String::from(kind),
+                edition: String::from(edition),
+                nested_int_test_files: test_files,
+            }
+        }
+
```
Comment (calebcartwright) [inline src/cargo-fmt/main.rs]: Running the tests directly against the `get_targets` [function](https://github.com/rust-lang/rustfmt/blob/a15e97f1e92b3621ab8c2f1585f27e8de583f74b/src/cargo-fmt/main.rs#L310) using the same sample projects under the `tests` directory. I wasn't originally planning on testing the new behavior this way, but it turns out to be rather tricky to instantiate some of the structs from `cargo_metadata` (no constructors, private struct fields, etc.) that would have been needed to test the functions that work with `cargo_metadata` structs directly (for example the `add_targets` function). 

I think this actually turns out to be a good thing though as these tests help validate the target determination process more holistically.

File: src/cargo-fmt/main.rs [hunk 6]
```diff
@@ -445,10 +505,59 @@ fn get_targets_with_hitlist(
...
+    }
+}
+
+// Returns a `Vec` containing `PathBuf`s of nested .rs files within subdirectories
+// of the `tests` directory for a given package.
+// https://github.com/rust-lang/rustfmt/issues/1820
+fn get_nested_integration_test_files(path: &Path, root_dir: &Path) -> Vec<PathBuf> {
+    let mut files = vec![];
+    if path.is_dir() {
+        for entry in fs::read_dir(path).expect(&format!(
```
Comment (topecongiro) [inline src/cargo-fmt/main.rs]: Please avoid calling `expect` here.

File: src/cargo-fmt/main.rs [hunk 7]
```diff
@@ -445,10 +505,59 @@ fn get_targets_with_hitlist(
...
+            None
+        };
+        targets.insert(Target::from_target(target, test_files));
+    }
+}
+
+// Returns a `Vec` containing `PathBuf`s of nested .rs files within subdirectories
+// of the `tests` directory for a given package.
+// https://github.com/rust-lang/rustfmt/issues/1820
+fn get_nested_integration_test_files(path: &Path, root_dir: &Path) -> Vec<PathBuf> {
```
Comment (topecongiro) [inline src/cargo-fmt/main.rs]: We can change the signature to `Option<Vec<PathBuf>>` and handle errors by transferring them into `None`.

File: src/cargo-fmt/main.rs [hunk 8]
```diff
@@ -445,10 +505,59 @@ fn get_targets_with_hitlist(
...
+    let mut files = vec![];
+    if path.is_dir() {
+        for entry in fs::read_dir(path).expect(&format!(
+            "couldn't read directory {}",
+            path.to_str().unwrap()
+        )) {
+            let entry = entry.expect("couldn't get nested `DirEntry` in tests");
```
Comment (topecongiro) [inline src/cargo-fmt/main.rs]: And this `expect` as well.

File: src/cargo-fmt/main.rs [hunk 9]
```diff
@@ -245,17 +259,27 @@ pub struct Target {
...
 
 impl Target {
-    pub fn from_target(target: &cargo_metadata::Target) -> Self {
+    pub fn from_target(
+        target: &cargo_metadata::Target,
+        nested_int_test_files: Option<Vec<PathBuf>>,
+    ) -> Self {
         let path = PathBuf::from(&target.src_path);
         let canonicalized = fs::canonicalize(&path).unwrap_or(path);
+        let test_files = match nested_int_test_files {
```
Comment (topecongiro) [inline src/cargo-fmt/main.rs]: Nit: `nested_int_test_files.unwrap_or(vec![])` ?

File: src/cargo-fmt/main.rs [hunk 10]
```diff
@@ -245,17 +259,27 @@ pub struct Target {
     kind: String,
     /// Rust edition for this target.
     edition: String,
+    // Rust files residing within subdirectories of the tests directory.
```
Comment (topecongiro) [inline src/cargo-fmt/main.rs]: We can make this as doc comment:

```suggestion
    /// Rust files residing within subdirectories of the tests directory.
```

File: src/cargo-fmt/main.rs (hunk 6 above)
Comment (calebcartwright) [reply src/cargo-fmt/main.rs]: 🤦‍♂ sorry, will do. I even made a [note](https://github.com/rust-lang/rustfmt/pull/3777#discussion_r321028028) to myself to remove them but forgot 😆

File: src/cargo-fmt/main.rs [hunk 11]
```diff
@@ -445,10 +502,65 @@ fn get_targets_with_hitlist(
...
+        targets.insert(Target::from_target(target, test_files));
+    }
+}
+
+// Returns a `Vec` containing `PathBuf`s of nested .rs files within subdirectories
+// of the `tests` directory for a given package.
+// https://github.com/rust-lang/rustfmt/issues/1820
+fn get_nested_integration_test_files(path: &Path, root_dir: &Path) -> Option<Vec<PathBuf>> {
+    let mut files = vec![];
+    if path.is_dir() {
```
[the remark this answers is not in the document]
Comment (calebcartwright) [reply src/cargo-fmt/main.rs]: > The indentations here are getting a bit too deep

Agreed! Will work on cleaning it up

[the remark this answers is not in the document]
Comment (calebcartwright) [reply src/cargo-fmt/main.rs]: I've cleaned up this function and IMO it looks a lot better now. Thanks for the suggested changes!

**Comment by calebcartwright** [comment]
I also made an udpate to fail if the user has included the `--include-nested-test-files` arg but we encounter an IO error while trying to discover those files. 

I feel like this is the correct/desired behavior, as we wouldn't want to silently ignore such file discovery errors and give users the (potentially) incorrect impression that those files are formatted correctly.

File: src/cargo-fmt/main.rs [hunk 12]
```diff
@@ -455,10 +512,70 @@ fn get_targets_with_hitlist(
...
+        }
+    }
+
+    Ok(())
+}
+
+// Returns a `Vec` containing `PathBuf`s of nested .rs files within subdirectories
+// of the `tests` directory for a given package.
+// https://github.com/rust-lang/rustfmt/issues/1820
+fn get_nested_integration_test_files(parent_path: &Path, root_dir: &Path) -> Option<Vec<PathBuf>> {
```
Comment (topecongiro) [inline src/cargo-fmt/main.rs]: Nit: the naming `parent_path` seems a bit odd, as it points to the actual target directory we are going to traverse.

Comment (calebcartwright) [reply src/cargo-fmt/main.rs]: It does now that you mention it. I'll revert it back

Comment (calebcartwright) [reply src/cargo-fmt/main.rs]: Done!

**Pull request merged**